### 모듈 불러오기

In [1]:
import pandas as pd
from pandas import Series, DataFrame
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns

##### 데이터 불러오기

In [2]:
import os
os.listdir("C:/Users/gun67/전력 프로젝트/2014")

['a. 2014_1_House1_Ch1_Combined_Data_New.csv',
 'b. 2014_2_House1_Ch1_Combined_Data_New.csv',
 'c. 2014_3_House1_Ch1_Combined_Data_New.csv',
 'd. 2014_4_House1_Ch1_Combined_Data_New.csv',
 'e. 2014_5_House1_Ch1_Combined_Data_New.csv',
 'f. 2014_6_House1_Ch1_Combined_Data_New.csv',
 'g. 2014_7_House1_Ch1_Combined_Data_New.csv',
 'h. 2014_8_House1_Ch1_Combined_Data_New.csv',
 'i. 2014_9_House1_Ch1_Combined_Data_New.csv',
 'j. 2014_10_House1_Ch1_Combined_Data_New.csv',
 'k. 2014_11_House1_Ch1_Combined_Data_New.csv',
 'l. 2014_12_House1_Ch1_Combined_Data_New.csv']

In [3]:
cd C:/Users/gun67/전력 프로젝트/2014

C:\Users\gun67\전력 프로젝트\2014


In [4]:
forders = os.listdir("C:/Users/gun67/전력 프로젝트/2014")
df_all_data = pd.DataFrame()

for files in forders :
    df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
    df_all_data = pd.concat([df_all_data, df], axis = 0)

C:\Users\gun67\AppData\Local\Temp\ipykernel_13944\2837981553.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\ipykernel_13944\2837981553.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\ipykernel_13944\2837981553.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Te

In [5]:
df_all_data

Unnamed: 0 Year-Month-Date       Time  Holiday Watt  Temp('C)  \
0               0      2014-01-01  [ 00:00 ]        1  570     9.000   
1               1      2014-01-01  [ 00:01 ]        1  603     8.967   
2               2      2014-01-01  [ 00:02 ]        1  579     8.933   
3               3      2014-01-01  [ 00:03 ]        1  576     8.900   
4               4      2014-01-01  [ 00:04 ]        1  602     8.867   
...           ...             ...        ...      ...  ...       ...   
44635       44635      2014-12-31  [ 23:55 ]        0  188     6.000   
44636       44636      2014-12-31  [ 23:56 ]        0  187     6.000   
44637       44637      2014-12-31  [ 23:57 ]        0  189     6.000   
44638       44638      2014-12-31  [ 23:58 ]        0  187     6.000   
44639       44639      2014-12-31  [ 23:59 ]        0  189     6.000   

       Humidity(%)  CPI  
0           76.000  1.9  
1           76.167  1.9  
2           76.333  1.9  
3           76.500  1.9  
4           76.667  1.9  
...            ...  ...  
44635       76.000  0.5  
44636       76.000  0.5  
44637       76.000  0.5  
44638       76.000  0.5  
44639       76.000  0.5  

[525600 rows x 8 columns]

### 불필요한 열 제거

In [6]:
df_all_data.drop(columns = ["Unnamed: 0", "CPI"], inplace = True)
df_all_data

Year-Month-Date       Time  Holiday Watt  Temp('C)  Humidity(%)
0          2014-01-01  [ 00:00 ]        1  570     9.000       76.000
1          2014-01-01  [ 00:01 ]        1  603     8.967       76.167
2          2014-01-01  [ 00:02 ]        1  579     8.933       76.333
3          2014-01-01  [ 00:03 ]        1  576     8.900       76.500
4          2014-01-01  [ 00:04 ]        1  602     8.867       76.667
...               ...        ...      ...  ...       ...          ...
44635      2014-12-31  [ 23:55 ]        0  188     6.000       76.000
44636      2014-12-31  [ 23:56 ]        0  187     6.000       76.000
44637      2014-12-31  [ 23:57 ]        0  189     6.000       76.000
44638      2014-12-31  [ 23:58 ]        0  187     6.000       76.000
44639      2014-12-31  [ 23:59 ]        0  189     6.000       76.000

[525600 rows x 6 columns]

### 인덱스 초기화

In [7]:
df_all_data.reset_index(drop = True, inplace = True)
df_all_data

Year-Month-Date       Time  Holiday Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]        1  570     9.000       76.000
1           2014-01-01  [ 00:01 ]        1  603     8.967       76.167
2           2014-01-01  [ 00:02 ]        1  579     8.933       76.333
3           2014-01-01  [ 00:03 ]        1  576     8.900       76.500
4           2014-01-01  [ 00:04 ]        1  602     8.867       76.667
...                ...        ...      ...  ...       ...          ...
525595      2014-12-31  [ 23:55 ]        0  188     6.000       76.000
525596      2014-12-31  [ 23:56 ]        0  187     6.000       76.000
525597      2014-12-31  [ 23:57 ]        0  189     6.000       76.000
525598      2014-12-31  [ 23:58 ]        0  187     6.000       76.000
525599      2014-12-31  [ 23:59 ]        0  189     6.000       76.000

[525600 rows x 6 columns]

### 데이터 타입 확인

In [8]:
print(df_all_data.dtypes)

Year-Month-Date     object
Time                object
Holiday              int64
Watt                object
Temp('C)           float64
Humidity(%)        float64
dtype: object


##### 숫자형 데이터로 표시되야 할 Watt 열에 잘못 기입된 데이터가 있다고 판단  
##### 데이터 타입을 숫자형으로 바꾸기 테스트 진행

In [9]:
df_all_data["Watt"] = pd.to_numeric(df_all_data["Watt"], errors = 'coerce')
print(df_all_data.dtypes)

Year-Month-Date     object
Time                object
Holiday              int64
Watt                 int64
Temp('C)           float64
Humidity(%)        float64
dtype: object


##### 다행히 Watt 열의 데이터 타입이 수치형으로 변경되었음을 알 수 있음

### 결손치 확인

In [10]:
print(df_all_data[df_all_data['Watt'] == 0])
print("-----------------------------------------")
print(df_all_data[df_all_data["Temp('C)"] == "NaN"])
print("-----------------------------------------")
print(df_all_data[(df_all_data["Humidity(%)"] < 0) | (df_all_data["Humidity(%)"] > 100)])

       Year-Month-Date       Time  Holiday  Watt  Temp('C)  Humidity(%)
88679       2014-03-03  [ 13:59 ]        0     0     8.000       87.000
88680       2014-03-03  [ 14:00 ]        0     0     8.000       87.000
88681       2014-03-03  [ 14:01 ]        0     0     8.000       86.800
88682       2014-03-03  [ 14:02 ]        0     0     8.000       86.600
88683       2014-03-03  [ 14:03 ]        0     0     8.000       86.400
...                ...        ...      ...   ...       ...          ...
290480      2014-07-21  [ 17:20 ]        0     0    24.667       58.333
290481      2014-07-21  [ 17:21 ]        0     0    24.700       58.200
290482      2014-07-21  [ 17:22 ]        0     0    24.733       58.067
290487      2014-07-21  [ 17:27 ]        0     0    24.900       57.400
290488      2014-07-21  [ 17:28 ]        0     0    24.933       57.267

[940 rows x 6 columns]
-----------------------------------------
Empty DataFrame
Columns: [Year-Month-Date, Time, Holiday, Watt, Temp('

##### Watt 열에서는 525600개의 레코드 중 940개, 약 0.18%의 결손치가 확인 됨  
##### 2013년도 데이터보다 결손치가 현저히 적지만, 데이터 구조의 통일성을 위해 똑같이 처리  
##### Temp('C) 열과 Humidicy(%) 열은 결손치가 없다고 판단

### 결손치 처리

In [11]:
# 전력값 결손치를 평균값으로 대체
data_watt = pd.DataFrame(df_all_data["Watt"])
print("결손치 대체 전 데이터")
print(data_watt)

# 0이 아닌 값을 가지는 데이터의 평균 계산
Watt_mean = data_watt[data_watt["Watt"] != 0]["Watt"].mean()

# 결손치를 평균값으로 대체
df_all_data.loc[data_watt["Watt"] == 0, "Watt"] = Watt_mean

print("\n결손치 대체 후 데이터")
print(df_all_data)

결손치 대체 전 데이터
        Watt
0        570
1        603
2        579
3        576
4        602
...      ...
525595   188
525596   187
525597   189
525598   187
525599   189

[525600 rows x 1 columns]

결손치 대체 후 데이터
       Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]        1  570.0     9.000       76.000
1           2014-01-01  [ 00:01 ]        1  603.0     8.967       76.167
2           2014-01-01  [ 00:02 ]        1  579.0     8.933       76.333
3           2014-01-01  [ 00:03 ]        1  576.0     8.900       76.500
4           2014-01-01  [ 00:04 ]        1  602.0     8.867       76.667
...                ...        ...      ...    ...       ...          ...
525595      2014-12-31  [ 23:55 ]        0  188.0     6.000       76.000
525596      2014-12-31  [ 23:56 ]        0  187.0     6.000       76.000
525597      2014-12-31  [ 23:57 ]        0  189.0     6.000       76.000
525598      2014-12-31  [ 23:58 ]        0  187.0     6.000 

C:\Users\gun67\AppData\Local\Temp\ipykernel_13944\280170690.py:10: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '439.0333663706019' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df_all_data.loc[data_watt["Watt"] == 0, "Watt"] = Watt_mean


### 이상치 확인

In [12]:
# Watt

# 통계적 방법으로 이상치 확인
# Z-점수 ; 각 데이터 포인트가 평균에서 몇 표준편차 떨어져 있는지 나타냄, 일반적으로 Z-점수가 3 이상이거나 -3 이하인 경우 이상치로 간주
# 사분위 범위 ; Q1(1사분위수)와 Q3(3사분위수) 사이의 범위를 계산하고, Q1 - 1.5 * IQR 보다 작거나 Q3 + 1.5 * IQR 보다 큰 값을 이상치로 간주

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Watt"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Watt"].quantile(0.25)
Q3 = df_all_data["Watt"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Watt"] < (Q1 - 1.5 * IQR)) | (df_all_data["Watt"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_data_watt = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_data_watt)

Z-scores :  0         0.229882
1         0.287806
2         0.245679
3         0.240414
4         0.286051
            ...   
525595    0.440632
525596    0.442387
525597    0.438876
525598    0.442387
525599    0.438876
Name: Watt, Length: 525600, dtype: float64
IQR :  215.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday    Watt  Temp('C)  Humidity(%)
110         2014-01-01  [ 01:50 ]        1  2636.0     7.000         85.0
111         2014-01-01  [ 01:51 ]        1  2666.0     7.000         85.2
112         2014-01-01  [ 01:52 ]        1  2657.0     7.000         85.4
113         2014-01-01  [ 01:53 ]        1  2662.0     7.000         85.6
114         2014-01-01  [ 01:54 ]        1  2740.0     7.000         85.8
...                ...        ...      ...     ...       ...          ...
523553      2014-12-30  [ 13:53 ]        0   833.0     7.000         66.0
523960      2014-12-30  [ 20:40 ]        0   810.0     3.667         83.0
523961      2014-12-30  [ 20:41 

In [13]:
# Temp('C)

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Temp('C)"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Temp('C)"].quantile(0.25)
Q3 = df_all_data["Temp('C)"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Temp('C)"] < (Q1 - 1.5 * IQR)) | (df_all_data["Temp('C)"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_df_Temp = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_df_Temp)

Z-scores :  0         0.782396
1         0.788709
2         0.795213
3         0.801526
4         0.807839
            ...   
525595    1.356303
525596    1.356303
525597    1.356303
525598    1.356303
525599    1.356303
Name: Temp('C), Length: 525600, dtype: float64
IQR :  8.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
285881      2014-07-18  [ 12:41 ]        0  256.0    29.100       55.033
285882      2014-07-18  [ 12:42 ]        0  256.0    29.200       54.400
285883      2014-07-18  [ 12:43 ]        0  252.0    29.300       53.767
285884      2014-07-18  [ 12:44 ]        0  254.0    29.400       53.133
285885      2014-07-18  [ 12:45 ]        0  253.0    29.500       52.500
...                ...        ...      ...    ...       ...          ...
286194      2014-07-18  [ 17:54 ]        0  546.0    29.200       39.000
286195      2014-07-18  [ 17:55 ]        0  550.0    29.167       39.167
286196      2014-07-18  [ 17:56 ]      

In [14]:
# humidity(%)

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Humidity(%)"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Humidity(%)"].quantile(0.25)
Q3 = df_all_data["Humidity(%)"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Humidity(%)"] < (Q1 - 1.5 * IQR)) | (df_all_data["Humidity(%)"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_df_humidity = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_df_humidity)

Z-scores :  0         0.174303
1         0.185711
2         0.197051
3         0.208459
4         0.219867
            ...   
525595    0.174303
525596    0.174303
525597    0.174303
525598    0.174303
525599    0.174303
Name: Humidity(%), Length: 525600, dtype: float64
IQR :  21.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
106048      2014-03-15  [ 15:28 ]        1  223.0    17.933       32.467
106049      2014-03-15  [ 15:29 ]        1  225.0    18.000       32.000
106050      2014-03-15  [ 15:30 ]        1  224.0    18.000       32.000
106051      2014-03-15  [ 15:31 ]        1  227.0    17.967       32.067
106052      2014-03-15  [ 15:32 ]        1  224.0    17.933       32.133
...                ...        ...      ...    ...       ...          ...
314879      2014-08-07  [ 15:59 ]        0  156.0    27.000       32.000
314880      2014-08-07  [ 16:00 ]        0  158.0    27.000       32.000
314881      2014-08-07  [ 16:01 ]  

### 이상치 처리

In [15]:
# Watt

# 이상치 대체 ; 평균, 중앙값 등으로 대체
# 이상치를 중앙값으로 대체
median_value = df_all_data["Watt"].median()
df_all_data.loc[(z_scores >= 3), "Watt"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]        1  570.0     9.000       76.000
1           2014-01-01  [ 00:01 ]        1  603.0     8.967       76.167
2           2014-01-01  [ 00:02 ]        1  579.0     8.933       76.333
3           2014-01-01  [ 00:03 ]        1  576.0     8.900       76.500
4           2014-01-01  [ 00:04 ]        1  602.0     8.867       76.667
...                ...        ...      ...    ...       ...          ...
525595      2014-12-31  [ 23:55 ]        0  188.0     6.000       76.000
525596      2014-12-31  [ 23:56 ]        0  187.0     6.000       76.000
525597      2014-12-31  [ 23:57 ]        0  189.0     6.000       76.000
525598      2014-12-31  [ 23:58 ]        0  187.0     6.000       76.000
525599      2014-12-31  [ 23:59 ]        0  189.0     6.000       76.000

[525600 rows x 6 columns]


In [16]:
# Temp

# 이상치를 중앙값으로 대체
median_value = df_all_data["Temp('C)"].median()
df_all_data.loc[(z_scores >= 3), "Temp('C)"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]        1  570.0     9.000       76.000
1           2014-01-01  [ 00:01 ]        1  603.0     8.967       76.167
2           2014-01-01  [ 00:02 ]        1  579.0     8.933       76.333
3           2014-01-01  [ 00:03 ]        1  576.0     8.900       76.500
4           2014-01-01  [ 00:04 ]        1  602.0     8.867       76.667
...                ...        ...      ...    ...       ...          ...
525595      2014-12-31  [ 23:55 ]        0  188.0     6.000       76.000
525596      2014-12-31  [ 23:56 ]        0  187.0     6.000       76.000
525597      2014-12-31  [ 23:57 ]        0  189.0     6.000       76.000
525598      2014-12-31  [ 23:58 ]        0  187.0     6.000       76.000
525599      2014-12-31  [ 23:59 ]        0  189.0     6.000       76.000

[525600 rows x 6 columns]


In [17]:
# Humidity

# 이상치를 중앙값으로 대체
median_value = df_all_data["Humidity(%)"].median()
df_all_data.loc[(z_scores >= 3), "Humidity(%)"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]        1  570.0     9.000       76.000
1           2014-01-01  [ 00:01 ]        1  603.0     8.967       76.167
2           2014-01-01  [ 00:02 ]        1  579.0     8.933       76.333
3           2014-01-01  [ 00:03 ]        1  576.0     8.900       76.500
4           2014-01-01  [ 00:04 ]        1  602.0     8.867       76.667
...                ...        ...      ...    ...       ...          ...
525595      2014-12-31  [ 23:55 ]        0  188.0     6.000       76.000
525596      2014-12-31  [ 23:56 ]        0  187.0     6.000       76.000
525597      2014-12-31  [ 23:57 ]        0  189.0     6.000       76.000
525598      2014-12-31  [ 23:58 ]        0  187.0     6.000       76.000
525599      2014-12-31  [ 23:59 ]        0  189.0     6.000       76.000

[525600 rows x 6 columns]


### 데이터셋 통합

In [18]:
Watt_N = pd.DataFrame(df_all_data["Watt"], columns = ["Watt"])
Temp_N = pd.DataFrame(df_all_data["Temp('C)"], columns = ["Temp('C)"])
Humidity_N = pd.DataFrame(df_all_data["Humidity(%)"], columns = ["Humidity(%)"])

In [19]:
df_all = pd.concat([df_all_data["Year-Month-Date"], df_all_data["Time"], Watt_N, Temp_N, Humidity_N], axis = 1)
df_all

Year-Month-Date       Time   Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]  570.0     9.000       76.000
1           2014-01-01  [ 00:01 ]  603.0     8.967       76.167
2           2014-01-01  [ 00:02 ]  579.0     8.933       76.333
3           2014-01-01  [ 00:03 ]  576.0     8.900       76.500
4           2014-01-01  [ 00:04 ]  602.0     8.867       76.667
...                ...        ...    ...       ...          ...
525595      2014-12-31  [ 23:55 ]  188.0     6.000       76.000
525596      2014-12-31  [ 23:56 ]  187.0     6.000       76.000
525597      2014-12-31  [ 23:57 ]  189.0     6.000       76.000
525598      2014-12-31  [ 23:58 ]  187.0     6.000       76.000
525599      2014-12-31  [ 23:59 ]  189.0     6.000       76.000

[525600 rows x 5 columns]

### 데이터 저장

In [20]:
df_all.to_csv("C:/Users/gun67/전력 프로젝트/preprocessing_data/2014_preprocessing_data", index = False, encoding = "utf-8-sig")

### 저장 데이터 확인

In [21]:
pd.read_csv("C:/Users/gun67/전력 프로젝트/preprocessing_data/2014_preprocessing_data", encoding = "utf-8-sig")

Year-Month-Date       Time   Watt  Temp('C)  Humidity(%)
0           2014-01-01  [ 00:00 ]  570.0     9.000       76.000
1           2014-01-01  [ 00:01 ]  603.0     8.967       76.167
2           2014-01-01  [ 00:02 ]  579.0     8.933       76.333
3           2014-01-01  [ 00:03 ]  576.0     8.900       76.500
4           2014-01-01  [ 00:04 ]  602.0     8.867       76.667
...                ...        ...    ...       ...          ...
525595      2014-12-31  [ 23:55 ]  188.0     6.000       76.000
525596      2014-12-31  [ 23:56 ]  187.0     6.000       76.000
525597      2014-12-31  [ 23:57 ]  189.0     6.000       76.000
525598      2014-12-31  [ 23:58 ]  187.0     6.000       76.000
525599      2014-12-31  [ 23:59 ]  189.0     6.000       76.000

[525600 rows x 5 columns]